# <u>**ACADEMY**</u> : Educational Systems Data Analysis

<details open>
<summary><u style="font-size: 2rem">Summary: </u></summary>

* <a href="#author" data-toc-modified-id="author" >Author</a>
* <a href="#introduction" data-toc-modified-id="introduction" >Introduction</a>
* <a href="#functions" data-toc-modified-id="functions" >Functions</a>
* <a href="#import-necessary-libraries" data-toc-modified-id="import-necessary-libraries" >Import Necessary Libraries</a>
* <a href="#load-the-data" data-toc-modified-id="load-the-data" >Load the Data</a>
* <a href="#initial-data-exploration" data-toc-modified-id="initial-data-exploration">Initial Data Exploration</a>
    * <a href="#stats-data" data-toc-modified-id="stats-data">1. EdStatsData</a>
* <a href="#data-cleaning" data-toc-modified-id="data-cleaning">Data Cleaning</a>



</details>

<a id='author'></a>

## Author
> Mohamed Ali EL HAMECH 

- @MasterCodeDevelop (Github)
- E-mail : master.code.develop@gmail.com
- Affiliation : ACADEMY #OpenClassRooms
<p><img align="left" src="https://github-readme-stats.vercel.app/api/top-langs?username=mastercodedevelop&show_icons=true&locale=en&layout=compact" alt="mastercodedevelop" />

<a id='introduction'></a>

## Introduction:
>In the rapidly evolving landscape of online education, ACADEMY, a burgeoning EdTech start-up, has been making significant strides by offering high-quality online training content tailored for high school and university students. As part of its strategic vision, ACADEMY is keenly exploring opportunities for international expansion, aiming to tap into markets with a robust educational framework and a potential clientele for its services.

_The objective of this analysis is twofold. Firstly, it seeks to delve into global educational data, sourced from the World Bank, to ascertain the viability and potential of various countries as prospective markets for ACADEMY. This dataset, curated by the "EdStats All Indicator Query" of the World Bank, boasts a comprehensive collection of over 4,000 international indicators. These indicators span a range of metrics, from access to education and graduation rates to insights about educators and educational expenditures._

_Secondly, this analysis aims to provide a clear, data-driven narrative that would aid ACADEMY's decision-makers in charting the company's international trajectory. By evaluating the quality of the dataset, understanding its breadth and depth, and extracting relevant insights, we hope to offer a roadmap that aligns with ACADEMY's mission and vision._

<a id='import-necessary-libraries'></a>

## Import Necessary Libraries
> Before you can work with the data, you need to import the necessary libraries.

In [161]:
import pandas as pd
import numpy as np

<a id='functions'></a>

## Functions

In [162]:
def missing_values_table(df):
    """
    Return a DataFrame displaying the count and percentage of missing values for each column.
    
    Parameters:
    - df (DataFrame): The input DataFrame to check for missing values.
    
    Returns:
    - DataFrame: A DataFrame with columns 'Missing Values' and 'Missing Percentages', sorted in descending order of missing percentages.
    """
    
    # Calculate missing values and their percentages
    missing_values = df.isnull().sum()
    missing_percentages = (df.isnull().mean() * 100).round(2)  # rounding to 2 decimal places for clarity
    
    # Create a DataFrame to display missing data info
    missing_data = pd.DataFrame({
        'Missing Values': missing_values, 
        'Missing Percentages (%)': missing_percentages
    })
    
    # Filter out columns with no missing values and sort by percentage
    missing_data = missing_data[missing_data['Missing Values'] > 0].sort_values(by='Missing Percentages (%)', ascending=False)
    
    return missing_data


<a id='load-the-data'></a>

## Load the Data
> Importing the dataset into the environment for analysis and exploration.

In [163]:
# These are the basic data variables we will be working with
COUNTRY_SERIES = pd.read_csv('./data/EdStatsCountry-Series.csv')
COUNTRY = pd.read_csv('./data/EdStatsCountry.csv')
DATA = pd.read_csv('./data/EdStatsData.csv')
FOOT_NOTE = pd.read_csv('./data/EdStatsFootNote.csv')
EdStatsSeries = pd.read_csv('./data/EdStatsSeries.csv')

# These are the data variables we will be working with and can be modified
country_series = COUNTRY_SERIES.copy()
country = COUNTRY.copy()
data = DATA.copy()
foot_note = FOOT_NOTE.copy()


<a id='initial-data-exploration'></a>

## Initial Data Exploration
> Diving into the dataset to understand its structure, content, and characteristics.

<a id='stats-data'></a>

### 1 - <u>EdStatsData :</u>

#### 1.1 - Preview of the First and Last Rows of the Dataset
> View of the dataset's start & end, providing a quick data structure glimpse.

In [164]:
print("First 5 rows of the dataset:")
DATA.head()

First 5 rows of the dataset:


Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        1970       1971       1972       1973       1974      1975  ...  2060  \
0        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
1        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
2        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
3        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
4  54.822121  54.894138  56.209438  57.267109  57.991138  59.36554  ...   NaN   

   2065  2070  2075  2080  2085  2090  2095  2100  Unnamed: 69  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
4   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  

[5 rows x 70 columns]

In [165]:
print("Last 5 rows of the dataset:")
print(DATA.tail())

Last 5 rows of the dataset:
       Country Name Country Code  \
886925     Zimbabwe          ZWE   
886926     Zimbabwe          ZWE   
886927     Zimbabwe          ZWE   
886928     Zimbabwe          ZWE   
886929     Zimbabwe          ZWE   

                                           Indicator Name  \
886925  Youth illiterate population, 15-24 years, male...   
886926  Youth literacy rate, population 15-24 years, b...   
886927  Youth literacy rate, population 15-24 years, f...   
886928  Youth literacy rate, population 15-24 years, g...   
886929  Youth literacy rate, population 15-24 years, m...   

              Indicator Code  1970  1971  1972  1973  1974  1975  ...  2060  \
886925      UIS.LP.AG15T24.M   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886926     SE.ADT.1524.LT.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886927  SE.ADT.1524.LT.FE.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886928  SE.ADT.1524.LT.FM.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   

> <u>Note:</u> By examining the first and last 5 lines, we notice a significant number of missing data, marked as 'NaN'.

#### 1.2 - General information about the DataFrame

In [166]:
# This includes the data type, number of non-zero values, etc.
print(DATA.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 886930 entries, 0 to 886929
Data columns (total 70 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Country Name    886930 non-null  object 
 1   Country Code    886930 non-null  object 
 2   Indicator Name  886930 non-null  object 
 3   Indicator Code  886930 non-null  object 
 4   1970            72288 non-null   float64
 5   1971            35537 non-null   float64
 6   1972            35619 non-null   float64
 7   1973            35545 non-null   float64
 8   1974            35730 non-null   float64
 9   1975            87306 non-null   float64
 10  1976            37483 non-null   float64
 11  1977            37574 non-null   float64
 12  1978            37576 non-null   float64
 13  1979            36809 non-null   float64
 14  1980            89122 non-null   float64
 15  1981            38777 non-null   float64
 16  1982            37511 non-null   float64
 17  1983      

In [167]:
# Display data types of each column
print(DATA.dtypes)

# Count columns by their data type
dtype_counts = DATA.dtypes.value_counts()
print("\nNumber of columns by data type:")
print(dtype_counts)


Country Name       object
Country Code       object
Indicator Name     object
Indicator Code     object
1970              float64
                   ...   
2085              float64
2090              float64
2095              float64
2100              float64
Unnamed: 69       float64
Length: 70, dtype: object

Number of columns by data type:
float64    66
object      4
dtype: int64


> <u>Note :</u>  While examining the data types of the columns, we observe that the first 4 columns are of type object, indicating they contain strings. Whereas columns with years as titles are of type float64, signifying they hold numerical data.

#### 1.3 - Columns

In [168]:
# Display columns names
print(DATA.columns)

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977', '1978',
       '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986', '1987',
       '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995', '1996',
       '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004', '2005',
       '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013', '2014',
       '2015', '2016', '2017', '2020', '2025', '2030', '2035', '2040', '2045',
       '2050', '2055', '2060', '2065', '2070', '2075', '2080', '2085', '2090',
       '2095', '2100', 'Unnamed: 69'],
      dtype='object')


> <u>Note :</u> When observing the columns, we first notice that the titles of the first 4 columns are words. Then, starting from the fifth column, the titles represent years, ranging from 1970 to 2100 in ascending order. From 1970 to 2017, the years increase by 1. There are no columns for the years between 2017 and 2020, while from 2020 to 2100, the years increase in intervals of 5. Lastly, there's a column named 'Unnamed: 69' which appears to be there by mistake.

#### 1.4 - Missing Values & Percentages
> Overview of missing data in columns, presented as counts & percentages

In [169]:
print("\nNumber of missing values for each column:")
DATA.isnull()


Number of missing values for each column:


Country Name  Country Code  Indicator Name  Indicator Code   1970  \
0              False         False           False           False   True   
1              False         False           False           False   True   
2              False         False           False           False   True   
3              False         False           False           False   True   
4              False         False           False           False  False   
...              ...           ...             ...             ...    ...   
886925         False         False           False           False   True   
886926         False         False           False           False   True   
886927         False         False           False           False   True   
886928         False         False           False           False   True   
886929         False         False           False           False   True   

         1971   1972   1973   1974   1975  ...  2060  2065  2070  2075  2080  \
0        True   True   True   True   True  ...  True  True  True  True  True   
1        True   True   True   True   True  ...  True  True  True  True  True   
2        True   True   True   True   True  ...  True  True  True  True  True   
3        True   True   True   True   True  ...  True  True  True  True  True   
4       False  False  False  False  False  ...  True  True  True  True  True   
...       ...    ...    ...    ...    ...  ...   ...   ...   ...   ...   ...   
886925   True   True   True   True   True  ...  True  True  True  True  True   
886926   True   True   True   True   True  ...  True  True  True  True  True   
886927   True   True   True   True   True  ...  True  True  True  True  True   
886928   True   True   True   True   True  ...  True  True  True  True  True   
886929   True   True   True   True   True  ...  True  True  True  True  True   

        2085  2090  2095  2100  Unnamed: 69  
0       True  True  True  True         True  
1       True  True  True  True         True  
2       True  True  True  True         True  
3       True  True  True  True         True  
4       True  True  True  True         True  
...      ...   ...   ...   ...          ...  
886925  True  True  True  True         True  
886926  True  True  True  True         True  
886927  True  True  True  True         True  
886928  True  True  True  True         True  
886929  True  True  True  True         True  

[886930 rows x 70 columns]

In [170]:
print(missing_values_table(DATA))

             Missing Values  Missing Percentages (%)
Unnamed: 69          886930                   100.00
2017                 886787                    99.98
2016                 870470                    98.14
1971                 851393                    95.99
1973                 851385                    95.99
...                     ...                      ...
2011                 740918                    83.54
2012                 739666                    83.40
2000                 710254                    80.08
2005                 702822                    79.24
2010                 644488                    72.67

[66 rows x 2 columns]


In [171]:
# Display missing data > 70%
missing_data_70 = missing_values_table(DATA)[missing_values_table(DATA)['Missing Percentages (%)'] > 70]
print("Display missing data > 70% :")
print(missing_data_70)
print( f"total columns that have missing data > 70% :", missing_data_70.index.size)

Display missing data > 70% :
             Missing Values  Missing Percentages (%)
Unnamed: 69          886930                   100.00
2017                 886787                    99.98
2016                 870470                    98.14
1971                 851393                    95.99
1973                 851385                    95.99
...                     ...                      ...
2011                 740918                    83.54
2012                 739666                    83.40
2000                 710254                    80.08
2005                 702822                    79.24
2010                 644488                    72.67

[66 rows x 2 columns]
total columns that have missing data > 70% : 66


In [172]:
# Display missing data > 90%
missing_data_90 = missing_values_table(DATA)[missing_values_table(DATA)['Missing Percentages (%)'] > 90]
print("Display missing data > 90% :")
print(missing_data_90)
print( f"total columns that have missing data > 90% :", missing_data_90.index.size)

Display missing data > 90% :
             Missing Values  Missing Percentages (%)
Unnamed: 69          886930                   100.00
2017                 886787                    99.98
2016                 870470                    98.14
1971                 851393                    95.99
1973                 851385                    95.99
1972                 851311                    95.98
1974                 851200                    95.97
1979                 850121                    95.85
1982                 849419                    95.77
1989                 849390                    95.77
1976                 849447                    95.77
1977                 849356                    95.76
1978                 849354                    95.76
1983                 848470                    95.66
1988                 848378                    95.65
1984                 848324                    95.65
1987                 848289                    95.64
1981             

> <u>Note :</u> Out of the 70 columns, a significant number show missing data: 44 columns have over 90% missing data, and 65 have more than 70% missing. Notably, the column "Unnamed: 69" is entirely missing data, at a full 100%.

#### 1.5 - Descriptive Statistics
> Displaying descriptive statistics to understand central tendencies, dispersion, and shape of the dataset's distribution.

In [173]:
DATA.describe()

1970          1971          1972          1973          1974  \
count  7.228800e+04  3.553700e+04  3.561900e+04  3.554500e+04  3.573000e+04   
mean   1.974772e+09  4.253638e+09  4.592365e+09  5.105006e+09  5.401493e+09   
std    1.211687e+11  1.804814e+11  1.914083e+11  2.059170e+11  2.112150e+11   
min   -1.435564e+00 -1.594625e+00 -3.056522e+00 -4.032582e+00 -4.213563e+00   
25%    8.900000e-01  8.853210e+00  9.240920e+00  9.595200e+00  9.861595e+00   
50%    6.317724e+00  6.316240e+01  6.655139e+01  6.969595e+01  7.087760e+01   
75%    6.251250e+01  5.655200e+04  5.863650e+04  6.202900e+04  6.383675e+04   
max    1.903929e+13  1.986457e+13  2.100916e+13  2.238367e+13  2.282991e+13   

               1975          1976          1977          1978          1979  \
count  8.730600e+04  3.748300e+04  3.757400e+04  3.757600e+04  3.680900e+04   
mean   2.314288e+09  5.731808e+09  6.124437e+09  6.671489e+09  7.436724e+09   
std    1.375059e+11  2.215546e+11  2.325489e+11  2.473986e+11  2.660957e+11   
min   -3.658569e+00 -2.950945e+00 -3.174870e+00 -3.558749e+00 -2.973612e+00   
25%    1.400000e+00  9.312615e+00  9.519913e+00  1.000000e+01  1.000000e+01   
50%    9.677420e+00  7.101590e+01  7.133326e+01  7.290512e+01  7.510173e+01   
75%    7.854163e+01  5.682800e+04  5.739175e+04  5.940425e+04  6.411500e+04   
max    2.300634e+13  2.424128e+13  2.521383e+13  2.622101e+13  2.730873e+13   

       ...          2060          2065          2070          2075  \
count  ...  5.143600e+04  5.143600e+04  5.143600e+04  5.143600e+04   
mean   ...  7.224868e+02  7.271290e+02  7.283779e+02  7.266484e+02   
std    ...  2.215845e+04  2.287990e+04  2.352338e+04  2.408149e+04   
min    ... -1.630000e+00 -1.440000e+00 -1.260000e+00 -1.090000e+00   
25%    ...  3.000000e-02  3.000000e-02  2.000000e-02  2.000000e-02   
50%    ...  2.300000e-01  2.300000e-01  2.300000e-01  2.300000e-01   
75%    ...  7.505000e+00  7.500000e+00  7.300000e+00  7.100000e+00   
max    ...  2.951569e+06  3.070879e+06  3.169711e+06  3.246239e+06   

               2080          2085          2090          2095          2100  \
count  5.143600e+04  5.143600e+04  5.143600e+04  5.143600e+04  5.143600e+04   
mean   7.228327e+02  7.176899e+02  7.113072e+02  7.034274e+02  6.940296e+02   
std    2.455897e+04  2.496587e+04  2.530183e+04  2.556069e+04  2.574189e+04   
min   -9.200000e-01 -7.800000e-01 -6.500000e-01 -5.500000e-01 -4.500000e-01   
25%    1.000000e-02  1.000000e-02  1.000000e-02  1.000000e-02  1.000000e-02   
50%    2.300000e-01  2.300000e-01  2.300000e-01  2.300000e-01  2.200000e-01   
75%    6.722500e+00  6.080000e+00  5.462500e+00  4.680000e+00  4.032500e+00   
max    3.301586e+06  3.337871e+06  3.354746e+06  3.351887e+06  3.330484e+06   

       Unnamed: 69  
count          0.0  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN  

[8 rows x 66 columns]

> <u>Note :</u> By examining the standard deviation and quantiles for each column, it's evident that the average (mean), minimum (min), and maximum (max) values show significant variability across the different columns.

#### 1.6 - Display duplicate rows

In [174]:
duplicated_rows = DATA.duplicated().sum()
print(f"total duplicate rows:  {duplicated_rows}")

total duplicate rows:  0


> <u>Note :</u> We can observe that there are no duplicated rows.

#### 1.7 - Unique Values Check
> Checking unique values for categorical columns to understand the different categories present.

In [175]:
for column in DATA.select_dtypes(include=['object']).columns:
    print(f"Unique values for {column}: {DATA[column].nunique()}")
    print(DATA[column].unique())
    print("\n")

Unique values for Country Name: 242
['Arab World' 'East Asia & Pacific'
 'East Asia & Pacific (excluding high income)' 'Euro area'
 'Europe & Central Asia' 'Europe & Central Asia (excluding high income)'
 'European Union' 'Heavily indebted poor countries (HIPC)' 'High income'
 'Latin America & Caribbean'
 'Latin America & Caribbean (excluding high income)'
 'Least developed countries: UN classification' 'Low & middle income'
 'Low income' 'Lower middle income' 'Middle East & North Africa'
 'Middle East & North Africa (excluding high income)' 'Middle income'
 'North America' 'OECD members' 'South Asia' 'Sub-Saharan Africa'
 'Sub-Saharan Africa (excluding high income)' 'Upper middle income'
 'World' 'Afghanistan' 'Albania' 'Algeria' 'American Samoa' 'Andorra'
 'Angola' 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Aruba' 'Australia'
 'Austria' 'Azerbaijan' 'Bahamas, The' 'Bahrain' 'Bangladesh' 'Barbados'
 'Belarus' 'Belgium' 'Belize' 'Benin' 'Bermuda' 'Bhutan' 'Bolivia'
 'Bosnia and Herze

> <u>Note: </u> We observe the presence of country names, regions, and indicator code

#### 1.8 - Dataset Dimensions & Unique Entries Overview
> - the number of rows 
> - the number of rows columns
> - unique number of countries 
> - unique number of indicators

In [176]:
print(f"Number of rows: {DATA.shape[0]}")
print(f"Number of columns {DATA.shape[1]}")
print(f"Nimber of unique countries: {DATA['Country Name'].nunique()}")
print(f"Number of unique indicators: {DATA['Indicator Name'].nunique()}")

Number of rows: 886930
Number of columns 70
Nimber of unique countries: 242
Number of unique indicators: 3665


> <u>Note :</u> The dataset, spanning 886 930 rows and 70 columns, encompasses 242 countries and 3,665 unique indicators. Its vast scope suggests in-depth information for each country and indicator, potentially over multiple years. The diversity of countries and indicators attests to a wealth of information suitable for varied analyses. This composition implies a global perspective with multiple data points per country.

#### 1.9 - <u>Notes</u> : 

The dataset, consisting of 886,930 rows and 70 columns, covers 242 countries with 3,665 unique indicators. The first four columns contain strings, while the subsequent ones, titled by years, hold numerical data. There's an annual progression from 1970 to 2017, a gap between 2017 and 2020, and then a five-year progression up to 2100. Many missing data points are observed, especially in the 'Unnamed: 69' column. Data variability is significant, and no rows are duplicated. The dataset provides a comprehensive perspective, rich for diverse analyses.

<a id='data-cleaning'></a>
## Data Cleaning
> Address any issues in the dataset to ensure its quality and reliability. This involves handling missing values, duplicates, and any inconsistencies.

### Remove Unnecessary Columns

In [177]:
# Display missing DATA == 100%
print(missing_values_table(DATA)[missing_values_table(DATA)['Missing Percentages (%)'] == 100])

             Missing Values  Missing Percentages (%)
Unnamed: 69          886930                    100.0


> <u>Remove "Unnamed: 69:"</u> Since you mentioned earlier that this column had 100% missing values, it would make sense to remove it.

In [178]:
# Drop the 'Unnamed: 69' column as it contains 100% missing values

# Calculate the missing values and their percentages for each column
missing_data = missing_values_table(data)

# Identify columns that have 100% missing values
cols_to_drop = missing_data[missing_data['Missing Percentages (%)'] == 100].index.tolist()

# Check if there are any columns to drop
# If you run this code a second time it gives an empty list, this is normal because the column will be deleted here we just have the cad of 'Unnamed: 69'
if cols_to_drop:
    data.drop(cols_to_drop, axis=1, inplace=True)
    plural = 's' if len(cols_to_drop) > 1 else ''
    print(f"{len(cols_to_drop)} column{plural} dropped because it{plural if len(cols_to_drop) == 1 else ' they'} had 100% missing values: {', '.join(cols_to_drop)}")
else:
    print("No columns were dropped as none had 100% missing values.")


1 column dropped because it had 100% missing values: Unnamed: 69


In [179]:
# Display missing data after removing the column 'Unnamed: 69'
print("Display missing data after removing the column")
print(missing_values_table(data))

Display missing data after removing the column
      Missing Values  Missing Percentages (%)
2017          886787                    99.98
2016          870470                    98.14
1973          851385                    95.99
1971          851393                    95.99
1972          851311                    95.98
...              ...                      ...
2011          740918                    83.54
2012          739666                    83.40
2000          710254                    80.08
2005          702822                    79.24
2010          644488                    72.67

[65 rows x 2 columns]


### Duplicated rows
> Remove duplicated rows if any

In [180]:
if duplicated_rows > 0:
    data.drop_duplicates(inplace=True)
    print("Duplicated rows have been removed.")
else:
    print("No duplicated rows found.")

No duplicated rows found.
